Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,495 @@
+      return;
+    }
+
+    _ACTS_ANY_VERBOSE(
+        "Copy construct (this=" << this << ") at: " << (void*)m_data.data());
+
+    m_handler = other.m_handler;
+    assert(m_handler != nullptr && "Handler is null");
+    copyConstruct(other);
+  }
+
+  AnyBase& operator=(const AnyBase& other) {
+    _ACTS_ANY_VERBOSE("Copy assign (this=" << this
+                                           << ") at: " << (void*)m_data.data());
+
+    if (m_handler == nullptr && other.m_handler == nullptr) {
+      // both are empty, noop
+      return *this;
+    }
+
+    if (m_handler == nullptr) {  // this object is empty
+      m_handler = other.m_handler;
+      assert(m_handler && "Handler is null");
+      copyConstruct(other);
+    } else {
+      // @TODO: Support assigning between different types
+      if (m_handler != other.m_handler) {
+        throw std::bad_any_cast{};
+      }
+      copy(other);
+    }
+    return *this;
+  }
+
+  AnyBase(AnyBase&& other) {
+    _ACTS_ANY_VERBOSE(
+        "Move construct (this=" << this << ") at: " << (void*)m_data.data());
+    if (m_handler == nullptr && other.m_handler == nullptr) {
+      // both are empty, noop
+      return;
+    }
+
+    m_handler = other.m_handler;
+    assert(m_handler != nullptr && "Handler is null");
+    moveConstruct(std::move(other));
+  }
+
+  AnyBase& operator=(AnyBase&& other) {
+    _ACTS_ANY_VERBOSE("Move assign (this=" << this
+                                           << ") at: " << (void*)m_data.data());
+    if (m_handler == nullptr && other.m_handler == nullptr) {
+      // both are empty, noop
+      return *this;
+    }
+
+    if (m_handler == nullptr) {  // this object is empty
+      m_handler = other.m_handler;
+      assert(m_handler && "Handler is null");
+      moveConstruct(std::move(other));
+    } else {
+      // @TODO: Support assigning between different types
+      if (m_handler != other.m_handler) {
+        throw std::bad_any_cast{};
+      }
+      move(std::move(other));
+    }
+    return *this;
+  }
+
+  operator bool() const { return m_handler != nullptr; }
+
+ private:
+  void* dataPtr() {
+    if (m_handler->heap) {
+      return *reinterpret_cast<void**>(m_data.data());
+    } else {
+      return reinterpret_cast<void*>(m_data.data());
+    }
+  }
+
+  void setDataPtr(void* ptr) { *reinterpret_cast<void**>(m_data.data()) = ptr; }
+
+  const void* dataPtr() const {
+    if (m_handler->heap) {
+      return *reinterpret_cast<void* const*>(m_data.data());
+    } else {
+      return reinterpret_cast<const void*>(m_data.data());
+    }
+  }
+
+  struct Handler {
+    void (*destroy)(void* ptr) = nullptr;
+    void (*moveConstruct)(void* from, void* to) = nullptr;
+    void (*move)(void* from, void* to) = nullptr;
+    void* (*copyConstruct)(const void* from, void* to) = nullptr;
+    void (*copy)(const void* from, void* to) = nullptr;
+    bool heap{false};
+  };
+
+  template <typename T>
+  static Handler* makeHandler() {
```

Comment:
Consider constifying this pointer's target, and the inner static as well while you're at it.